bgpd/bgp_linkstate_tlv.c

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/* BGP Link-State TLV Serializer/Deserializer
+ * Copyright 2023 6WIND S.A.
+ */
+
+#include <zebra.h>
+
+#include "bgpd/bgpd.h"
+#include "bgpd/bgp_route.h"
+#include "bgpd/bgp_debug.h"
+#include "bgpd/bgp_errors.h"
+#include "bgpd/bgp_linkstate_tlv.h"
+
+
+static uint16_t pnt_decode16(uint8_t **pnt)
+{
+	uint16_t data;
+
+	*pnt = ptr_get_be16(*pnt, &data);
+
+	return data;
+}
+
+int bgp_nlri_parse_linkstate(struct peer *peer, struct attr *attr,
+			     struct bgp_nlri *packet, int withdraw)
+{
+	uint8_t *pnt;
+	uint8_t *lim;
+	afi_t afi;
+	safi_t safi;
+	uint16_t length = 0;
+	struct prefix p;
+
+	/* Start processing the NLRI - there may be multiple in the MP_REACH */
+	pnt = packet->nlri;
+	lim = pnt + packet->length;
+	afi = packet->afi;
+	safi = packet->safi;
+
+	for (; pnt < lim; pnt += length) {
+		/* Clear prefix structure. */
+		memset(&p, 0, sizeof(p));
+
+		/* All linkstate NLRI begin with NRLI type and length. */
+		if (pnt + 4 > lim)
+			return BGP_NLRI_PARSE_ERROR_PACKET_OVERFLOW;
+
+		p.u.prefix_linkstate.nlri_type = pnt_decode16(&pnt);
+		length = pnt_decode16(&pnt);
+		/* When packet overflow occur return immediately. */
+		if (pnt + length > lim) {
+			flog_err(
+				EC_BGP_LINKSTATE_PACKET,
+				"Link-State NLRI length inconsistent (size %u seen)",
+				length);
+			return BGP_NLRI_PARSE_ERROR_PACKET_OVERFLOW;
+		}
+		p.family = AF_LINKSTATE;
+
+		p.u.prefix_linkstate.ptr = (uintptr_t)pnt;
```

Comment:
I have a doubt on pnt.
pnt points in peer->curr, but I curr is a stream that is dependent of the incoming flow.
while we are on the same thread, I think it is ok to use that buffer, but if we are in a new job, we may face the risk of losing data. am I right?

Reply:
bgp_update() work on a copy of the memory zone
```
#4  0x00007f3fad0f9325 in prefix_copy (udest=..., usrc=...) at lib/prefix.c:393
#5  0x00007f3fad125337 in route_node_get (table=0x5571ecc5f890, pu=...) at lib/table.c:274
#6  0x00005571ead5f52f in bgp_node_get (table=0x5571ecc5f850, p=0x7ffd33e7b130) at ./bgpd/bgp_table.h:221
#7  0x00005571ead63a54 in bgp_afi_node_get (table=0x5571ecc5f850, afi=AFI_LINKSTATE, safi=SAFI_LINKSTATE, p=0x7ffd33e7b130, prd=0x0) at bgpd/bgp_route.c:146
#8  0x00005571ead6d23c in bgp_update (peer=0x7f3faa5ee010, p=0x7ffd33e7b130, addpath_id=0, attr=0x7ffd33e7b240, afi=AFI_LINKSTATE, safi=SAFI_LINKSTATE, type=9, sub_type=0, prd=0x0, label=0x0, num_labels=0, soft_reconfig=0, evpn=0x0)
    at bgpd/bgp_route.c:4059
#9  0x00005571ead302db in bgp_nlri_parse_linkstate (peer=0x7f3faa5ee010, attr=0x7ffd33e7b240, packet=0x7ffd33e7b210, withdraw=0) at bgpd/bgp_linkstate_tlv.c:292
#10 0x00005571ead4fa3c in bgp_nlri_parse (peer=0x7f3faa5ee010, attr=0x7ffd33e7b240, packet=0x7ffd33e7b210, mp_withdraw=false) at bgpd/bgp_packet.c:353
#11 0x00005571ead535ae in bgp_update_receive (peer=0x7f3faa5ee010, size=341) at bgpd/bgp_packet.c:2030
#12 0x00005571ead55cf6 in bgp_process_packet (thread=0x7ffd33e7b560) at bgpd/bgp_packet.c:2940
#13 0x00007f3fad12de90 in event_call (thread=0x7ffd33e7b560) at lib/event.c:1995
#14 0x00007f3fad0c285b in frr_run (master=0x5571ec47adc0) at lib/libfrr.c:1185
#15 0x00005571eaccac8d in main (argc=7, argv=0x7ffd33e7b7b8) at bgpd/bgp_main.c:505
```